transcom/mymove
# Kim mb 1045 payment request errors

## Description

For payment request creation:
 - adds a ValidationError 422 response to the yaml, and makes some of the descriptions more specific
 - ~Returns `verrs` from the PaymentRequestCreator to differentiate from server errors to use return a 422 Unprocessable Entity~ removed verrs and wrapped/unwrapped verrs instead
 - if a db query within the service returns no records, it now uses a custom error to eventually return a 400 instead of a 500
- adds some tests to the handler to check for the custom error and the verrs

## Setup

In Postman, try different scenarios for a POST of a payment request in the Prime Environment.  Add json in the body of the request (be sure to set the type to json).
Try different scenarios and see what response messages and status you get.  You may need to run a GET in ghc office environment for move-task-orders and use an moveTaskOrderID in the below json. For each error response you should get a body object that includes a title, detail, and instance.

ex. 
```
{
    "detail": "Invalid payment request: params Body is nil",
    "instance": "00000000-0000-0000-0000-000000000000",
    "title": "Unhandled SQL error encountered"
}
```


1) successful:
```
{
    "isFinal": false,
    "moveTaskOrderID": "8ce7926d-b774-4832-b3e4-c0b56043b878",
    "serviceItems": [
        {
            "id": "9db1bf43-0964-44ff-8384-3297951f6781",
            "params": [
                {
                    "key": "WeightEstimated",
                    "value": "3254"
                },
                {
                    "key": "RequestedPickupDate",
                    "value": "2020-04-20"
                }
            ]
        }
    ]
}
```

2) send a nil body: (you will get a 400)

3) send json without an MTO ID: (you will get a 400)

4) send non-existent MTO ID (I change a digit in the existing one): (you will get a 404)

5) send a non-existent MTO Service Item ID: (you will get a 404)

6) send a non-existent MTO service item param key: (you will get  a 400)

7) send a non-existent MTO service item param key ID: (you will get  a 400)

8) send an MTO ID with a bad format (I add a digit): based on the handler code, I expected a 400, but you get a 422 with a body that says 
```
{
    "code": 601,
    "message": "moveTaskOrderID in body must be of type uuid: \"8ce7926d-b774-4832-b3e4-c0b56043b8789\""
}
```
Apparently this is the error caught by Swagger before it even hits the handler for a bad uuid format.

9) the model validations for a payment request don't get hit because of earlier checks, but to manually test that you get a 422 and correct error message, comment out the call of the uniqueIdentifier (currently `payment_request_creator.go` L196).  You should get validation errors for two invalid fields: `payment_request_number` and `sequence_number`

## Reviewer Notes

I don't have any new tests for the service, though seems like there should be.  I couldn't figure out how to re-create a validation error.  The handler does a check for a bad MTO ID, which also catches a non-existent MTO ID, so the model validation to check that an MTO ID is present is never hit (which is partly why I couldn't create a test in the service).  I'm wondering if that should be reconfigured in some way so that we hit the validation.

The AC states "correctly formatted error body for all errors".  We are now returning a formatted body for validation errors according to the [backend errors doc](https://github.com/transcom/mymove/wiki/handle-backend-errors).  I'm not sure if other errors should also have similar types of bodies.

## Code Review Verification Steps

 [Querying the Database Safely](https://github.com/transcom/mymove/tree/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-1045) for this change
* [this article](tbd) explains more about the approach used.
